torch/csrc/jit/passes/peephole_list_idioms.cpp

```diff
@@ -21,17 +22,312 @@ c10::optional<size_t> normalizeIndex(int64_t index, size_t len) {
   }
 }
 
+// Refine from Value -> len of list
+// TODO: vector may be faster
+using ListRefinement = std::unordered_map<Value*, int64_t>;
+
+// The intersection of the refinements is the Value* which are in both
+// refinements and are refined to the same length
+ListRefinement intersectRefinements(
+    const ListRefinement& ref1,
+    const ListRefinement& ref2) {
+  ListRefinement out;
+  for (const auto& pair : ref1) {
+    auto val2 = ref2.find(pair.first);
+    if (val2 != ref2.end() && val2->second == pair.second) {
+      out[pair.first] = pair.second;
+    }
+  }
+  return out;
+}
+
+// To union, just take all refinements from both inputs. We do not need to worry
+// about len refinements disagreeing because a path like `if len(x) == 4 and
+// len(x) == 5` will never be taken
+ListRefinement unionRefinements(
+    const ListRefinement& ref1,
+    const ListRefinement& ref2) {
+  ListRefinement out;
+  for (const auto& pair : ref1) {
+    out[pair.first] = pair.second;
+  }
+  for (const auto& pair : ref2) {
+    out[pair.first] = pair.second;
+  }
+  return out;
+}
+
+// Represents the refinement information that can be carried on a boolean
+struct BoolRefinements {
```

Comment:
I think adding some examples of `TrueRefinement`, `FalseRefinement`, and `BoolRefine` w/ both `true_refine` and `false_refine` would make it easier to understand.

Reply:
Happy for naming suggestions.... We're associating Refinements with the boolean's true value, and different Refinements for the boolean's false value.

Reply:
One other slightly odd thing about `BoolRefinement` in comparison to `ListRefinement` is that the latter is a map from Value to some info we know about that value. But the former is itself a data we store in some map. I wonder if it would be more consistent if we called that outer map a `BoolRefinement`. Maybe what we have now is perfectly fine too, it's just something difficult to wrap one's head around.

Reply:
Not all refinements are associated with a boolean value. 
If you have something like
```
assert len(x) == 4
print(len(x))
```

After the assertion there are refinements present in the block that are not mapped to a particular value. So I'm not sure it makes sense to add a mapping of (BoolValue) to the core data structure, and would prefer the extra indirection of Value* -> pair(refinements) on the class itself.

Reply:
Yeah, that's fine. What I was pointing out is that we use the word "refinement" in one situation to describe info about all values (list values), and in the other situation about a single value, and that seems inconsistent.